### PR TITLE
[No CI] [DO NOT MERGE] Dispatcher Lab

### PR DESCRIPTION
**Pull request by SherlockNoMad**
### (gdb) break structured_mul_out::impl

**Before: 22 frames from Python Binding to CPU Kernel**
image

**After: 4 frames from Python Binding to CPU Kernel**
image

### Profiling
```
import torch
with torch.profiler.profile() as prof:
    a = torch.empty(2)
    b = torch.empty(2)
    c = torch.mul(a, b)
print(str(prof.events()))
```

**Before**
---------------  ------------  ------------  ------------  ------------  ------------  ------------  
           Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
    aten::empty         0.14%      41.000us         0.14%      41.000us      41.000us             1  
    aten::empty         0.00%       1.000us         0.00%       1.000us       1.000us             1  
      aten::mul        99.86%      29.180ms        99.86%      29.180ms      29.180ms             1  
Self CPU time total: 29.222ms

**After**
---------------  ------------  ------------  ------------  ------------  ------------  ------------  
           Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
    aten::empty        88.89%      64.000us        88.89%      64.000us      64.000us             1  
    aten::empty        11.11%       8.000us        11.11%       8.000us       8.000us             1  

Self CPU time total: 72.000us

### Benchmark
```
import torch
from torch.utils.benchmark import Timer

t = Timer(
    stmt="torch.mul(a, b)",
    setup="a=torch.empty(2);b=torch.empty(2)").timeit()
print(t)
```

**Before**
<torch.utils.benchmark.utils.common.Measurement object at 0x7fe8223e2a30>
torch.mul(a, b)
setup: a=torch.empty(2);b=torch.empty(2)
  2.26 us
  1 measurement, 1000000 runs , 1 thread

**After**
<torch.utils.benchmark.utils.common.Measurement object at 0x7fc42a808f10>
torch.mul(a, b)
setup: a=torch.empty(2);b=torch.empty(2)
  1.84 us
  1 measurement, 1000000 runs , 1 thread

### CODE REVIEW

File: aten/src/ATen/templates/Functions.cpp [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 #include <ATen/Functions.h>
 #include <ATen/Utils.h>
+#include <Aten/NativeFunctions.h>
```
Comment (SherlockNoMad) [inline aten/src/ATen/templates/Functions.cpp:5]: ATen!!!!!

File: aten/src/ATen/templates/TensorMethods.cpp [hunk 2]
```diff
@@ -28,4 +28,104 @@ namespace at {
  AT_FORALL_SCALAR_TYPES_WITH_COMPLEX(DEFINE_ITEM)
  #undef DEFINE_ITEM
 
- } //namespace at
+} //namespace at
+
+#include <Aten/NativeFunctions.h>
```
Comment (SherlockNoMad) [inline aten/src/ATen/templates/TensorMethods.cpp:33]: ATen!!!!

File: torch/csrc/lazy/core/shape_inference.h [hunk 3]
```diff
@@ -64,6 +64,9 @@ TORCH_API std::vector<Shape> compute_shape_sum(const at::Tensor & self, c10::opt
 TORCH_API std::vector<Shape> compute_shape__to_copy(const at::Tensor & self, c10::optional<at::ScalarType> dtype, c10::optional<at::Layout> layout, c10::optional<at::Device> device, c10::optional<bool> pin_memory, bool non_blocking, c10::optional<at::MemoryFormat> memory_format);
 TORCH_API std::vector<Shape> compute_shape_trace(const at::Tensor & self);
 TORCH_API std::vector<Shape> compute_shape_zero_(at::Tensor & self);
+TORCH_API std::vector<Shape> compute_shape_mul(const at::Tensor & self, const at::Tensor & other);
+TORCH_API std::vector<Shape> compute_shape_mul_(at::Tensor & self, const at::Tensor & other);
+TORCH_API std::vector<Shape> compute_shape_mul_out(at::Tensor& out, const at::Tensor& self, const at::Tensor& other);
```
Comment (bdhirsh) [inline torch/csrc/lazy/core/shape_inference.h:69]: Nice catch- looks like removing `mul` also breaks LazyTensor now - I'll mention this in the wiki page (basically, we now lose "meta tensor" support for the `mul` function, so LazyTensor expects it to be manually implemented here instead).

**Comment by bdhirsh** [comment]
This looks good to me! And nice benchmarks!

**Pull request closed without merging**